### PR TITLE
[lcms] Update to 2.16

### LINKED ISSUE #35835: [lcms] update to 2.16
### Library name

lcms

### New version number

2.16

### Other information that may be useful (release notes, etc...)

https://github.com/mm2/Little-CMS/releases/tag/lcms2.16

> lcms2-2.16 is a featured release
>
>Changes
>
>    New: import .CUBE files as RGB device links
    New: Read/Write MHC2 tags for Windows GPU access
    New: Support for UTF8 on multi-localized Unicode functions
    New: Support for OkLab color space, built-in and formatter.
    Improved: floating point transforms float -> integer are now honored as float
    Improved: MSYS2, mingw is now supported
    Improved: preferred CMM, platform and creator now survives profile edition.
    Fixed: tificc now can deal with Lab TIFF
    Fixed: code can now be compiled by a C++17 compiler, "register" keyword use detected at compile time.
    Fixed: Reverted PostScript creation that corrupted some interpreters.
    Bug fixing

### PR DESCRIPTION
**Pull request by LilyWangLL**
Fixes #35835, update `lcms` to 2.16, switch use `meson` to build it, and remove the old patches.
Usage passed on `x64-windows`.

- [x] Changes comply with the [maintainer guide](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/contributing/maintainer-guide.md)
- [x] SHA512s are updated for each updated download
- [ ] ~~The "supports" clause reflects platforms that may be fixed by this new version~~
- [ ] ~~Any fixed [CI baseline](https://github.com/microsoft/vcpkg/blob/master/scripts/ci.baseline.txt) entries are removed from that file.~~
- [x] Any patches that are no longer applied are deleted from the port's directory.
- [x] The version database is fixed by rerunning `./vcpkg x-add-version --all` and committing the result.
- [x] Only one version is added to each modified port's versions file.

### CODE REVIEW

File: ports/lcms/portfile.cmake [hunk 1]
```diff
@@ -1,36 +1,34 @@
...
-file(WRITE "${CURRENT_PACKAGES_DIR}/share/lcms/lcms-config.cmake" [[
-include(CMakeFindDependencyMacro)
-find_dependency(lcms2 CONFIG)
-include(${CMAKE_CURRENT_LIST_DIR}/lcms-targets.cmake)
-]])
 
-file(INSTALL "${SOURCE_PATH}/COPYING" DESTINATION "${CURRENT_PACKAGES_DIR}/share/${PORT}" RENAME copyright)
-file(INSTALL "${CMAKE_CURRENT_LIST_DIR}/usage" DESTINATION "${CURRENT_PACKAGES_DIR}/share/${PORT}")
+if(VCPKG_TARGET_IS_LINUX)
+    vcpkg_replace_string("${CURRENT_PACKAGES_DIR}/lib/pkgconfig/lcms2.pc" [[Cflags: "-I${includedir}" -pthread]] [[Cflags: "-I${includedir}"]])
```
Comment (dg0yt) [inline ports/lcms/portfile.cmake:27]: Why do you want to remove `-pthread`? It is in `Cflags` for several other ports. If openimageio doesn't handle it correctly (`-D-pthread`), shouldn't the fix be in that port? Or is this an artifact from `FindPkgConfig.cmake`?

Comment (LilyWangLL) [reply ports/lcms/portfile.cmake]: I reverted this change and fix in `openimageio`.

**Review by Cheney-W — DISMISSED** [review]

File: ports/lcms/portfile.cmake [hunk 2]
```diff
@@ -1,36 +1,27 @@
-vcpkg_minimum_required(VERSION 2022-10-12) # for ${VERSION}
-
-if(VCPKG_LIBRARY_LINKAGE STREQUAL dynamic)
-    set(SHARED_LIBRARY_PATCH "fix-shared-library.patch")
+set(SHARED_PATCHES )
+if(VCPKG_LIBRARY_LINKAGE STREQUAL "dynamic")
+    list(APPEND SHARED_PATCHES shared.patch)
 endif()
```
Comment (dg0yt) [inline ports/lcms/portfile.cmake:4]: Can this be turned into a patch which is always applied? It is easy to miss when updating from x64-linux.

**Review by dg0yt — COMMENTED** [review]
The old vcpkg cmake config is removed without mitigation for existing users. Bad UX.

The openimageio change should be integrated in, and follow the patterns in, `fix-dependencies.patch`.

File: ports/lcms/fix-shared-library.patch [hunk 3]
```diff
@@ -1,12 +1,12 @@
...
 +++ b/include/lcms2.h
-@@ -227,6 +227,7 @@ typedef int                  cmsBool;
+@@ -230,6 +230,7 @@ typedef int                  cmsBool;
  
  
  // Calling convention -- this is hardly platform and compiler dependent
 +#define CMS_DLL
- #ifdef CMS_IS_WINDOWS_
+ #if defined(CMS_IS_WINDOWS_) && !defined(__GNUC__)
  #  if defined(CMS_DLL) || defined(CMS_DLL_BUILD)
```
Comment (dg0yt) [inline ports/lcms/fix-shared-library.patch:11]: We should really get rid of this conditional patch.  
For static linkage, neither `CMS_DLL` nor `CMS_DLL_BUILD` should be defined.  
For dynamic linkage, `CMS_DLL_BUILD` should come the build system when building,  
and `defined(CMS_DLL)` should replaced with `1` after installation (aka "burn-in library linkage").

**Comment by dg0yt** [comment]
@LilyWangLL What happens here? Repo removed or hidden?

**Comment by LilyWangLL** [comment]
> What happens here? Repo removed or hidden?

Ah...I made a mistake and deleted the wrong branch. T-T

**Comment by LilyWangLL** [comment]
Reopen.

**Pull request closed without merging**